packages/server/src/adapters/lambda/index.ts

```diff
@@ -0,0 +1,149 @@
+import type { HTTPResponse } from '../../http/internals/types';
+import { AnyRouter, inferRouterContext } from '../../router';
+import type {
+  APIGatewayEvent,
+  APIGatewayResult,
+  AWSLambdaOptions,
+} from './utils';
+import {
+  UNKNOWN_PAYLOAD_FORMAT_VERSION,
+  isPayloadV1,
+  isPayloadV2,
+} from './utils';
+
+export type { CreateLambdaContextOptions, AWSLambdaOptions } from './utils';
+
+function lambdaEventToHTTPRequest(event: APIGatewayEvent): HTTPRequest {
+  const query = new URLSearchParams();
+  for (const [key, value] of Object.entries(
+    event.queryStringParameters ?? {},
+  )) {
+    if (value) {
+      query.append(key, value);
+    }
+  }
+
+  return {
+    method: getHTTPMethod(event),
+    query: query,
+    headers: event.headers,
+    body: event.body,
+  };
+}
+
+function getHTTPMethod(event: APIGatewayEvent) {
+  if (isPayloadV1(event)) {
+    return event.httpMethod;
+  } else if (isPayloadV2(event)) {
+    return event.requestContext.http.method;
+  } else {
+    throw UNKNOWN_PAYLOAD_FORMAT_VERSION;
+  }
+}
+function getPath(event: APIGatewayEvent) {
+  if (isPayloadV1(event)) {
+    return event.path.slice(1);
+  } else if (isPayloadV2(event)) {
+    return event.rawPath.slice(1);
+  } else {
+    throw UNKNOWN_PAYLOAD_FORMAT_VERSION;
+  }
+}
+function transformHeaders(headers: HTTPHeaders): APIGatewayResult['headers'] {
+  const obj: APIGatewayResult['headers'] = {};
+
+  for (const [key, value] of Object.entries(headers)) {
+    if (typeof value === 'undefined') {
+      continue;
+    }
+    obj[key] = Array.isArray(value) ? value.join(',') : value;
+  }
+  return obj;
+}
+function tRPCOutputToAPIGatewayOutput<
+  TEvent extends APIGatewayEvent,
+  TResult extends APIGatewayResult,
+>(event: TEvent, response: HTTPResponse): TResult {
+  if (isPayloadV1(event)) {
+    const resp: APIGatewayProxyResult = {
+      statusCode: response.status,
+      body: response.body ?? '',
+      headers: transformHeaders(response.headers ?? {}),
+    };
+    return resp as TResult;
+  } else if (isPayloadV2(event)) {
+    const resp: APIGatewayProxyStructuredResultV2 = {
+      statusCode: response.status,
+      body: response.body ?? undefined,
+      headers: transformHeaders(response.headers ?? {}),
+    };
+    return resp as TResult;
+  } else {
+    throw UNKNOWN_PAYLOAD_FORMAT_VERSION;
+  }
+}
+
+/** Will check the createContext of the TRouter and get the parameter of event.
+ * @internal
+ **/
+type DeduceAPIGWEventType<
+  TRouter extends AnyRouter,
+  TEvent extends APIGatewayEvent,
+> = AWSLambdaOptions<TRouter, TEvent>['createContext'] extends NonNullable<
+  AWSLambdaOptions<TRouter, TEvent>['createContext']
+>
+  ? Parameters<AWSLambdaOptions<TRouter, TEvent>['createContext']>[0]['event']
+  : APIGatewayEvent;
+
+/** 1:1 mapping of v1 or v2 input events, deduces which is which.
+ * @internal
+ **/
+type DeduceAPIGWReturnType<T> = T extends APIGatewayProxyEvent
```

Comment:
Minor: I think a better name would be 
```suggestion
type APIGWResult<T> = T extends APIGatewayProxyEvent
```

(Deduce is fine just seemed unnecessary - main thing is that it’s not technically a function return type, but a result type, which is an Aws rather than typescript concept)

Reply:
Maybe `inferAPIGWResult`? We use `infer` in other places like this.